WordPress/gutenberg
# Support Hot Module Replacement (HMR) for CSS

## What?

Part of https://github.com/WordPress/gutenberg/discussions/36142.

Add Hot Module Replacement for CSS. This is an experimental development-only feature.

## Why?

HMR for JavaScript is a little bit more complicated and supporting CSS also helps pave the way too.

Hot Module Replacement (or hot reload) is a bundler (webpack) feature that replaces the module upon saving the files without reloading the page. This PR adds support for CSS HMR so that updating CSS files locally automatically updates the stylesheets on the page in development mode. This helps developers iterate more quickly with faster visual feedback.

## How?

Due to how Gutenberg works, this is a bit complicated but I'll try my best explaining it 😅.

To understand how HMR works, we must understand the current way of building CSS in Gutenberg. When running `npm run dev`, there are multiple steps to complete to update the CSS on the page:

```mermaid
flowchart LR
    A(packages/*/src/style.scss) -->|SASS compiler / postCSS| B(packages/*/build-style/style.css)
    B -->|webpack| C(build/*/style.css)
```

1. `build-worker` builds the SASS source files to CSS files to each package's `build-style` folder. These files will be published to the npm registry.
2. webpack then copies those build files to the root `build` directory for the page to consume.

We don't change the first step in this PR to simplify and minimize the scope (but we could for future optimizations).

Now, let's move on to how HMR works in webpack. The official doc has a great [overview](https://webpack.js.org/concepts/hot-module-replacement/) of the technical background. The difference between the recommended setup and our setup are:

1. We don't import styles in our JavaScript code.
2. The styles are not registered by webpack, but instead managed by the PHP backend.

The first difference is relatively easy to fix if we just unshift the style entries for each JavaScript entry.

```js
module.exports = {
  entry: {
    'edit-post': {
-      import: './packages/edit-post',
+      import: [ './packages/edit-post/build-style/style.css', './packages/edit-post' ],
    }
  },
};
```

This tells webpack to load the styles in each JavaScript entry so that they can be processed by webpack and inject the HMR runtime. We want to keep the JavaScript file at the end so that they'll be exported to the `window.wp` property.

The second difference is important because we don't want to change the order of the styles to prevent breakage. This means we can't use [`style-loader`](https://webpack.js.org/loaders/style-loader/), which is the most popular way to inject and hot reload CSS files at the time.

Instead, we create a separate custom loader for CSS files that includes the HMR code to replace injected styles on the fly. I first tried out [`restyle-loader`](https://v4.webpack.js.org/loaders/restyle-loader/) which mostly did what we want but it lacks maintenance and doesn't work for iframed styles. [`css-hmr-loader`](https://github.com/WordPress/gutenberg/pull/64444/files#diff-5a8e5062e60ddef19260db7ba2a38a96c909ccc18243715f6ce73ace292ad5b8) is refactor that largely based on `restyle-loader` and `mini-css-extract-plugin` that suits our needs. The code uses the [HMR runtime API](https://webpack.js.org/api/hot-module-replacement/) to listen and react to code changes.

The next step is to start a dev server that listens to the HMR event on the client. [`webpack-dev-server`](https://webpack.js.org/guides/hot-module-replacement/#enabling-hmr) provides everything we need, we just need to provide a manual entry for the HMR runtime because we have multiple entry points. The dev server will listen on port `8887` which matches the port used in blocks development.

Setting the dev server up also means we no longer treat the file system as the source of truth, but rather the hot-enabled in-memory dev server. Any request to the original URL (`http://localhost:8888/wp-content/plugins/gutenberg/build/*`) will need to be redirected to the dev server (`http://localhost:8887/build/*`) for HMR to work. We achieve this by changing the implementation of `gutenberg_url` in `client-assets.php`. I'm not sure if that's enough, if not then we could also go a bit further and inject a custom `.htaccess` file.

As a bonus, but also pretty essential, is that we also want to continue to support running e2e tests in development mode. Since there's no difference from the PHP backend perspective to know whether it's loading a production build or a hot-enabled development mode, the most reliable way is to check whether a `hot.js` entry point is created in the build folder. If it exists then we assume it's running a development build, redirecting from `localhost:8888` to `localhost:8887`.

I believe this covers most of the implementation in this PR. Feel free to ask any further questions or if you spot any bugs!

The same implementation could and should be able to be used for JavaScript HMR too. But I think that's better left to another PR to minimize the scope 😅.

## Testing Instructions




1. Run `npm run dev` (run `rm -rf build` beforehand if you want to test it from scratch.)
2. Visit any editor and notice the WDS message in the console.
3. Edit some `.css` files and see if they reflect themselves on the page without a reload.
4. Try different setups and builds like the testing environment `localhost:8889` and production builds `npm run build`. Expect they still work correctly.

## Screenshots or screencast 

https://github.com/user-attachments/assets/6ffbee80-976a-4f45-b2cf-9238612000cd
